File: 前缀和/1590_使数组能被P整除/sol.js
/**
 * 定理一： 给定正整数x,y,z,p,如果 y mod p = x, 则（y-z）mod p = 0 等价于 z mod p = x.
 * 定理二： 给定正整数x,y,z,p （y-z）mod p = x 等价于 z mod p = (y-x)mod p.
 */

/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minSubarray = function (nums, p) {
  const n = nums.length;
  // 前缀和
  const s = new Array(n + 1).fill(0);
  for (let i = 0; i < n; i++) {
    s[i + 1] = s[i] + nums[i];
  }
  let x = s[s.length - 1] % p;
  if (x === 0) return 0;
  let ans = n;
  const map = new Map();
  let j = -n;
  for (let i = 0; i <= n; i++) {
    map.set(s[i] % p, i);
    if (map.has((s[i] - x) % p)) {
      j = map.get((s[i] - x) % p);
    }
    ans = Math.min(ans, i - j);
  }
  return ans < n ? ans : -1;
};

let nums = [11, 2, 5, 7, 8, 9],
  p = 10;
console.log(minSubarray(nums, p));

File: 前缀和/1590_使数组能被P整除/sol.py
from typing import List
from itertools import accumulate

class Solution:
    def minSubArray(self, nums: List[int], p:int)->int:
        s = list(accumulate(nums, initial = 0))
        x = s[-1] % p
        if x == 0: return 0 # 移除空子数组（这行可以不要）

        ans = n = len(nums)
        last = {}
        for i, v in enumerate(s):
            last[v % p] = i
            j = last.get((v-x) % p, -n) # 如果不存在，-n 可以保证 i-j >= n
            ans = min(ans, i-j)
        return ans if ans < n else -1

Sol = Solution()
res = Sol.minSubArray([11,2,5,7,8,9], 10)
print(res)